apps/anomalist/detectors.py

```diff
@@ -0,0 +1,193 @@
+from typing import Dict, List, Literal, Optional
+
+import numpy as np
+import pandas as pd
+import plotly.express as px
+import structlog
+from owid.datautils.dataframes import map_series
+
+from etl import grapher_model as gm
+from etl.data_helpers.misc import bard
+
+log = structlog.get_logger()
+
+# Name of index columns for dataframe.
+INDEX_COLUMNS = ["entity_id", "year"]
+
+# Define anomaly types.
+ANOMALY_TYPE = Literal["upgrade_change", "time_change", "upgrade_missing"]
+
+
+def estimate_bard_epsilon(series: pd.Series) -> float:
+    # Make all values positive, and ignore zeros.
+    positive_values = abs(series.dropna())
+    # Ignore zeros, since they can lead to epsilon being zero, hence allowing division by zero in BARD.
+    positive_values = positive_values.loc[positive_values > 0]
+    # Estimate epsilon as the absolute range of values divided by 10.
+    # eps = (positive_values.max() - positive_values.min()) / 10
+    # Instead of just taking maximum and minimum, take 95th percentile and 5th percentile.
+    eps = (positive_values.quantile(0.95) - positive_values.quantile(0.05)) / 10
+
+    return eps
+
+
+def get_long_format_score_df(df_score: pd.DataFrame) -> pd.DataFrame:
+    # Create a reduced score dataframe.
+    df_score_long = df_score.melt(
+        id_vars=["entity_id", "year"], var_name="variable_id", value_name="anomaly_score"
+    ).fillna(0)
+    # For now, keep only the latest year affected for each country-indicator.
+    df_score_long = (
+        df_score_long.sort_values("anomaly_score", ascending=False)
+        .drop_duplicates(subset=["variable_id", "entity_id"], keep="first")
+        .reset_index(drop=True)
+    )
+
+    return df_score_long
+
+
+class AnomalyDetector:
+    anomaly_type: str
+
+    def __init__(
+        self,
+        variable_ids: List[int],
+        df: pd.DataFrame,
+        metadata: Dict[int, gm.Variable],
+        variable_mapping: Dict[int, int],
+        entity_id_to_name: Dict[int, str],
+    ) -> None:
+        self.variable_ids = variable_ids
+        self.variable_mapping = variable_mapping
+
+        # Initialize data objects.
+        self.df = df
+        self.metadata = metadata
+        self.entity_id_to_name = entity_id_to_name
```

Comment:
Do we need to convert ids to names in every detector? We could just convert entity ids to names in `anomaly_detection` and all detectors would work with them. Or is there a benefit in passing them as ids?

Reply:
I suppose we can just use entity names everywhere. I thought sticking to entity ids made sense, just because we are using indicator ids everywhere too. But I can change that to use only entity names.